### PR TITLE
Add LLaVa-1.6, bis

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [LLaVa-1.6](https://llava-vl.github.io/blog/2024-01-30-llava-next/) (also called LLaVa-NeXT) as a standalone model rather than modifying the existing modeling_llava.py. As discussed in #29012 this might be better.

To do:

- [x] convert all checkpoints once PR is approved, along with corresponding integration test, which will also fix the failing documentation test.

### CODE REVIEW

**Comment by amyeroberts** [comment]
Thanks for opening @NielsRogge! I can see some model related tests are failing - ping me for review once they're all passing

**Comment by NielsRogge** [comment]
@amyeroberts I've fixed all tests, only the documentation test is failing because the official checkpoint hasn't been uploaded yet.

File: src/transformers/models/llava_next/modeling_llava_next.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+
+                inputs_embeds, attention_mask, labels, position_ids = self._merge_input_ids_with_image_features(
+                    image_features, inputs_embeds, input_ids, attention_mask, labels
+                )
+                if labels is None:
+                    labels = torch.full_like(attention_mask, self.config.ignore_index).to(torch.long)
+            else:
+                # In case input_ids.shape[1] == 1 & pixel_values==None & past_key_values != None, we are in the case of
+                # generation with cache
+                if past_key_values is not None and pixel_values is not None and input_ids.shape[1] == 1:
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/modeling_llava_next.py:550]: What if any of these conditions aren't true - what situation are we in? If it's not valid, we should raise an error. For example, can I call this method directly with `input_embeds` if `input_ids` is None?

```suggestion
            elif past_key_values is not None and pixel_values is not None and input_ids.shape[1] == 1:
                # In case input_ids.shape[1] == 1 & pixel_values==None & past_key_values != None, we are in the case of
                # generation with cache
```

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+        patch_size (`int`):
+            The size of each patch.
+        input_data_format (`ChannelDimension` or `str`):
+            The channel dimension format of the input image.
+
+    Returns:
+        list: A list of np.array representing the patches.
+    """
+    patches = []
+    height, width = image.shape[:-1] if input_data_format == ChannelDimension.LAST else image.shape[1:]
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:108]:
```suggestion
    height, width = get_image_size(image, input_data_format)
```

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+
+
+class LlavaNextImageProcessor(BaseImageProcessor):
+    r"""
+    Constructs a LLaVa-NeXT image processor. Based on [`CLIPImageProcessor`] with incorporation of additional techniques
+    for processing high resolution images as explained in the [LLaVa paper](https://arxiv.org/abs/2310.03744).
+
+    Args:
+        aspect_ratio_setting (`str`, *optional*, defaults to `"anyres"`):
+            The aspect ratio setting to use. Can be "clip" (as in CLIP), "pad" (as in LLaVa 1.5) or "anyres" (as in LLaVa 1.6).
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:145]: * I'm not sure it make sense to support the "pad" option now we have a new model 
* Under what cases should "clip" be used?

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+            `do_resize` in the `preprocess` method.
+        size (`Dict[str, int]` *optional*, defaults to `{"shortest_edge": 224}`):
+            Size of the image after resizing. The shortest edge of the image is resized to size["shortest_edge"], with
+            the longest edge resized to keep the input aspect ratio. Can be overridden by `size` in the `preprocess`
+            method.
+        image_grid_pinpoints (`List` *optional*, defaults to `[[672, 336], [336, 672], [672, 672], [336, 1008], [1008, 336]]`):
+            A list of possible resolutions to use for processing high resolution images. The best resolution is selected
+            based on the original size of the image. Can be overridden by `image_grid_pinpoints` in the `preprocess`
+            method.
+        resample (`PILImageResampling`, *optional*, defaults to `Resampling.BICUBIC`):
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:158]: Double checking this matches the original implementation

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+
+        scale_w = target_width / original_width
+        scale_h = target_height / original_height
+
+        if scale_w < scale_h:
+            new_width = target_width
+            new_height = min(math.ceil(original_height * scale_w), target_height)
+        else:
+            new_height = target_height
+            new_width = min(math.ceil(original_width * scale_h), target_width)
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:424]: As this logic is exactly the same as in `_resize_for_patching` we can abstract it out to a utility function something like `_get_patch_output_size(original_size, target_resolution)`

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+        image_size = get_image_size(image, channel_dim=input_data_format)
+        best_resolution = select_best_resolution(image_size, possible_resolutions)
+        resized_image = self._resize_for_patching(
+            image, best_resolution, resample=resample, input_data_format=input_data_format
+        )
+        padded_image = self._pad_for_patching(resized_image, best_resolution, input_data_format=input_data_format)
+
+        patches = divide_to_patches(padded_image, patch_size=patch_size, input_data_format=input_data_format)
+
+        # make sure that all patches use the input data format
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:479]: This comment isn't clear - the logic below is making sure all patches are in the `data_format` format, not 'using' the input data format

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+        )
+
+        image_patches = [resized_original_image] + patches
+
+        return image_patches
+
+    def preprocess(
+        self,
+        images: ImageInput,
+        aspect_ratio_setting=None,
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/image_processing_llava_next.py:500]:
```suggestion
        aspect_ratio_setting: Optional[str] = None,
```

File: tests/models/llava_next/test_modeling_llava_next.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            "type_vocab_size": 16,
+            "type_sequence_label_size": 2,
+            "initializer_range": 0.02,
+            "num_labels": 3,
+            "num_choices": 4,
+            "pad_token_id": 0,
+        },
+        is_training=True,
+        vision_config={
+            "batch_size": 12,
```
Comment (amyeroberts) [inline tests/models/llava_next/test_modeling_llava_next.py:79]: Why do we need batch_size here?

File: tests/models/llava_next/test_modeling_llava_next.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
+
+    @unittest.skip(reason="Feedforward chunking is not yet supported")
+    def test_feed_forward_chunking(self):
+        pass
+
+    @unittest.skip(reason="CPU offload is not yet supported")
+    def test_cpu_offload(self):
+        pass
```
Comment (amyeroberts) [inline tests/models/llava_next/test_modeling_llava_next.py:216]: This is funny - if we have `_no_split_modules` set then it should be supported. Otherwise the model will automatically be skipped for this test, so we shouldn't need it

File: tests/models/llava_next/test_modeling_llava_next.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+
+
+@require_torch
+class LlavaNextForConditionalGenerationModelTest(ModelTesterMixin, unittest.TestCase):
+    """
+    Model tester for `LlavaNextForConditionalGeneration`.
+    """
+
+    all_model_classes = (LlavaNextForConditionalGeneration,) if is_torch_available() else ()
+    fx_compatible = False
```
Comment (amyeroberts) [inline tests/models/llava_next/test_modeling_llava_next.py:167]: Was this tested?

File: tests/models/llava_next/test_modeling_llava_next.py [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+        EXPECTED_DECODED_TEXT = "USER:  \nWhat are the things I should be cautious about when I visit this place?\nASSISTANT: When visiting this place, which is a pier or dock extending over a body of water, there are a few things to be cautious about. First, be aware of the weather conditions, as sudden changes in weather can make the pier unsafe to walk on. Second, be mindful of the water depth and any potential hazards, such as submerged rocks or debris, that could cause accidents or injuries. Additionally, be cautious of the presence of wildlife, such as birds or fish, and avoid disturbing their natural habitats. Lastly, be aware of any local regulations or guidelines for the use of the pier, as some areas may be restricted or prohibited for certain activities."  # fmt: skip
+
+        self.assertEqual(
+            processor.decode(output[0], skip_special_tokens=True),
+            EXPECTED_DECODED_TEXT,
+        )
+
+    @slow
+    @require_bitsandbytes
+    def test_small_model_integration_test_llama_batched(self):
```
Comment (amyeroberts) [inline tests/models/llava_next/test_modeling_llava_next.py:276]: mistral, based on the checkpoint 

```suggestion
    def test_small_model_integration_test_mistral_batched(self):
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this as a new architecture - it's a lot cleaner this way!

File: src/transformers/models/llava_next/processing_llava_next.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+              `None`).
+            - **pixel_values** -- Pixel values to be fed to a model. Returned when `images` is not `None`.
+        """
+        if images is not None:
+            image_inputs = self.image_processor(images, return_tensors=return_tensors)
+        else:
+            image_inputs = {}
+        text_inputs = self.tokenizer(
+            text, return_tensors=return_tensors, padding=padding, truncation=truncation, max_length=max_length
+        )
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/processing_llava_next.py:111]: Done

File: src/transformers/models/llava_next/configuration_llava_next.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        self.vocab_size = self.vocab_size
+
+        self.text_config = text_config
+
+        if isinstance(self.text_config, dict):
+            text_config["model_type"] = text_config["model_type"] if "model_type" in text_config else "llama"
+            self.text_config = CONFIG_MAPPING[text_config["model_type"]](**text_config)
+            self.vocab_size = self.text_config.vocab_size
+        elif text_config is None:
+            self.text_config = CONFIG_MAPPING["llama"]()
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/configuration_llava_next.py]: Good point, I've removed `vocab_size` since this can be taken from `text_config`.

cc @ArthurZucker wondering why you added `vocab_size` as a separate attribute for llava

File: src/transformers/models/llava_next/modeling_llava_next.py [hunk 14]
```diff
@@ -0,0 +1,684 @@
...
+    @property
+    def _supports_sdpa(self):
+        """
+        Retrieve language_model's attribute to check whether the model supports
+        SDPA or not.
+        """
+        return self.language_model._supports_sdpa
+
+
+LLAVA_NEXT_INPUTS_DOCSTRING = r"""
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/modeling_llava_next.py:225]: Added those

File: docs/source/en/model_doc/llava_next.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+rendered properly in your Markdown viewer.
+
+-->
+
+# LLaVA-NeXT
+
+## Overview
+
+The LLaVA-NeXT model was proposed in [LLaVA-NeXT: Improved reasoning, OCR, and world knowledge](https://llava-vl.github.io/blog/2024-01-30-llava-next/) by Haotian Liu, Chunyuan Li, Yuheng Li, Bo Li, Yuanhan Zhang, Sheng Shen, Yong Jae Lee. LLaVa-NeXT (also called LLaVa-1.6) improves upon [LLaVa](llava) by increasing the input image resolution and training on an improved visual instruction tuning dataset to improve OCR and common sense reasoning.
+
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply docs/source/en/model_doc/llava_next.md:22]: The model architecture is the same, the only differs is the increased image resolution. I've added a figure to the documentation for clarity

File: tests/models/llava_next/test_modeling_llava_next.py (hunk 10 above)
Comment (NielsRogge) [reply tests/models/llava_next/test_modeling_llava_next.py]: This actually works, so removing this for both llava-next and llava

File: src/transformers/models/llava_next/image_processing_llava_next.py
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/image_processing_llava_next.py:1]: Tests are added

File: src/transformers/models/llava_next/image_processing_llava_next.py (hunk 4 above)
Comment (NielsRogge) [reply src/transformers/models/llava_next/image_processing_llava_next.py:175]: Yes, original implementation also uses BICUBIC interpolation, which is asserted in the conversion script [here](https://github.com/NielsRogge/transformers/blob/add_llava_1_6_bis/src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py#L174)

**Comment by NielsRogge** [comment]
Thanks for your review, addressed all comments and uploaded all checkpoints to a [collection](https://huggingface.co/collections/llava-hf/llava-next-65f75c4afac77fd37dbbe6cf).

The documentation test fails, seems like the machine crashed due to the size of the model (cc @ydshieh).

File: tests/models/llava_next/test_modeling_llava_next.py (hunk 9 above)
Comment (NielsRogge) [reply tests/models/llava_next/test_modeling_llava_next.py:215]: Pinging @younesbelkada here regarding how `_no_split_modules` should be defined for llava/llava-next.

**Comment by ydshieh** [comment]
> Thanks for your review, addressed all comments and uploaded all checkpoints to a [collection](https://huggingface.co/collections/llava-hf/llava-next-65f75c4afac77fd37dbbe6cf).
> 
> The documentation test fails, seems like the machine crashed due to the size of the model (cc @ydshieh).

If you can make use lower resource (anyway like fp16, 4-bit etc) that would be great.
If nothing could avoid the crash, just put the file to `utils/not_doctested.txt`

**Comment by amyeroberts** [comment]
@NielsRogge Please review the PR before asking for review. There's files which should not be included as part of the PR here

**Comment by NielsRogge** [comment]
Apologies, scripts are removed

File: src/transformers/models/llava_next/configuration_llava_next.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            Custom vision config or dict
+        text_config (`Union[AutoConfig, dict]`, *optional*):
+            The config object of the text backbone. Can be any of `LlamaConfig` or `MistralConfig`.
+        ignore_index (`int`, *optional*, defaults to -100):
+            The ignore index for the loss function.
+        image_token_index (`int`, *optional*, defaults to 32000):
+            The image token index to encode the image prompt.
+        projector_hidden_act (`str`, *optional*, defaults to `"gelu"`):
+            The activation function used by the multimodal projector.
+        vision_feature_select_strategy (`str`, *optional*, defaults to `"default"`):
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/llava_next/configuration_llava_next.py:49]: OK - can you add an explanation here about what "default" versus "full" means?

File: src/transformers/models/llava_next/image_processing_llava_next.py (hunk 3 above)
Comment (amyeroberts) [reply src/transformers/models/llava_next/image_processing_llava_next.py:162]: To be addressed

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this. It's almost ready, but some more coverage is needed for tests. In particular, the generation tests need to be added for the model

File: src/transformers/models/llava_next/configuration_llava_next.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/configuration_llava_next.py:49]: Added

File: src/transformers/models/llava_next/image_processing_llava_next.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/llava_next/image_processing_llava_next.py:162]: "clip" isn't really used (that's the default way of how CLIP preprocesses images), only "pad" is used (for llava 1.5) and "anyres" (for llava 1.6).

[People have reported](https://huggingface.co/llava-hf/llava-1.5-7b-hf/discussions/18) that llava 1.5-hf has the wrong preprocessing settings ("clip"), so they could use the "pad" setting if they want one-on-one matching results:

File: tests/models/llava_next/test_image_processor_llava_next.py
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/llava_next/test_image_processor_llava_next.py:1]: Yes, we test both `aspect_ratio_setting = "clip"` and `aspect_ratio_setting = "anyres"`

File: src/transformers/models/llava_next/image_processing_llava_next.py [hunk 17]
```diff
@@ -0,0 +1,704 @@
...
+                    data_format=data_format,
+                    input_data_format=input_data_format,
+                )[0]
+
+            new_images.append(pixel_values)
+
+        if aspect_ratio_setting == "anyres":
+            data = {"pixel_values": new_images, "image_sizes": image_sizes}
+        elif aspect_ratio_setting in ["clip", "pad"]:
+            data = {"pixel_values": new_images}
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava_next/image_processing_llava_next.py:702]: That's a good point, actually `LlavaNextForConditionalGeneration` expects the `image_sizes` argument.

The other 2 settings ("clip" and "pad") are only there for people that want to have the correct image processing settings for `LlavaForConditionalGeneration`.

File: src/transformers/models/llava/configuration_llava.py [hunk 18]
```diff
@@ -116,15 +121,17 @@ def __init__(
...
             text_config["model_type"] = text_config["model_type"] if "model_type" in text_config else "llama"
-            self.text_config = CONFIG_MAPPING[text_config["model_type"]](**text_config)
-            self.vocab_size = self.text_config.vocab_size
+            text_config = CONFIG_MAPPING[text_config["model_type"]](**text_config)
+            vocab_size = text_config.vocab_size
         elif text_config is None:
-            self.text_config = CONFIG_MAPPING["llama"]()
+            text_config = CONFIG_MAPPING["llama"]()
+
+        self.vocab_size = vocab_size
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/llava/configuration_llava.py]: Deprecated `vocab_size` as it can be taken from `text_config`

**Comment by NielsRogge** [comment]
Regarding your comment regarding the `aspect_ratio_setting` of [LlavaNextImageProcessor](url) -> actually only `aspect_ratio_setting = "anyres"` is used for llava 1.6 (`LlavaNextForConditionalGeneration`).

The other 2 settings ("clip" and "pad") are there to be used for `LlavaForConditionalGeneration` (the previous generation). This one currently uses CLIPImageProcessor, which is technically incorrect, it should use the "pad" setting. But I assume we can't simply replace the image processor of `LlavaForConditionalGeneration` by a new one.

Should I keep the `aspect_ratio_setting`, or remove it and only support "anyres" by default?

**Comment by amyeroberts** [comment]
> This one currently uses CLIPImageProcessor, which is technically incorrect, it should use the "pad" setting of LlavaNextImageProcessor. But I assume we can't simply replace the image processor of LlavaForConditionalGeneration by a new one (cc @ArthurZucker).

I don't quite understand this statement. "This one" is referring to `LlavaForConditionalGeneration` or `LlavaNextForConditionalGeneration`? 

> But I assume we can't simply replace the image processor of LlavaForConditionalGeneration by a new one (cc @ArthurZucker).

Again, not sure which object you're talking about replacing, but if there's an incorrect implementation it would be better to fix  it, possibly with some kind of deprecation cycle / legacy behaviour flag. If the issue is with the pad method, I don't see why we need to completely replace the image processor, instead of updating `pad`

> Should I keep the aspect_ratio_setting, or remove it and only support "anyres" by default?

We should remove the `aspect_ratio_setting` flag entirely

It doesn't make sense to have an image processor for a specific model, which is preparing inputs for a different model which already has its own image processor.

**Comment by NielsRogge** [comment]
Ok, I've removed the flag as it leads to confusion. Feel free to approve the PR, if possible.

File: src/transformers/models/llava_next/image_processing_llava_next.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/llava_next/image_processing_llava_next.py]: Removed this flag

File: src/transformers/models/llava_next/modeling_llava_next.py [hunk 19]
```diff
@@ -0,0 +1,690 @@
...
```
Comment (amyeroberts) [inline src/transformers/models/llava_next/modeling_llava_next.py:225]: docstring doesn't match signature

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating! 

PR can't be merged as-is as it introduces breaking changes to existing models: llava and vip llava

File: tests/models/llava/test_modeling_llava.py [hunk 20]
```diff
@@ -159,9 +159,7 @@ class LlavaForConditionalGenerationModelTest(ModelTesterMixin, unittest.TestCase
 
     all_model_classes = (LlavaForConditionalGeneration,) if is_torch_available() else ()
     pipeline_model_mapping = {"image-to-text": LlavaForConditionalGeneration} if is_torch_available() else {}
-    fx_compatible = False
     test_pruning = False
-    test_resize_embeddings = True
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/llava/test_modeling_llava.py:164]: Because `test_resize_embeddings` is set to `True` by default in test_modeling_common.py

File: tests/models/vipllava/test_modeling_vipllava.py [hunk 21]
```diff
@@ -185,6 +186,188 @@ def test_training_gradient_checkpointing_use_reentrant(self):
...
+            model_not_tied = model_class(config)
+            if model_not_tied.get_output_embeddings() is None:
+                continue
+
+            config_tied = copy.deepcopy(config)
+            config_tied.torchscript = False
+            model_tied = model_class(config_tied)
+            params_tied = list(model_tied.parameters())
+            # Check that the embedding layer and decoding layer are the same in size and in value
+            # self.assertTrue(check_same_values(embeddings, decoding))
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/vipllava/test_modeling_vipllava.py:347]: This is copied from the [common test](https://github.com/huggingface/transformers/blob/66ce9593fdb8e340df546ddd0774eb444f17a12c/tests/test_modeling_common.py#L1923)

File: tests/models/llava/test_modeling_llava.py [hunk 22]
```diff
@@ -186,6 +184,188 @@ def test_training_gradient_checkpointing_use_reentrant(self):
     def test_training_gradient_checkpointing_use_reentrant_false(self):
         pass
 
+    # Copied from tests.test_modeling_common.ModelTesterMixin.test_resize_tokens_embeddings with config.vocab_size->config.text_config.vocab_size
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/llava/test_modeling_llava.py:187]: I've tried that, but that doesn't work. The test fails with:
```python
# Check that resizing the token embeddings with a larger vocab size increases the model's vocab size
            model_embed = model.resize_token_embeddings(model_vocab_size + 10)
>           self.assertEqual(model.config.vocab_size, model_vocab_size + 10)
E           AssertionError: 99 != 109
```
This is because when calling the resize_token_embeddings method, we are changing `config.text_config.vocab_size`, whereas `config.vocab_size` remains unchanged.

File: tests/models/llava_next/test_modeling_llava_next.py [hunk 23]
```diff
@@ -0,0 +1,477 @@
...
+            model_not_tied = model_class(config)
+            if model_not_tied.get_output_embeddings() is None:
+                continue
+
+            config_tied = copy.deepcopy(config)
+            config_tied.torchscript = False
+            model_tied = model_class(config_tied)
+            params_tied = list(model_tied.parameters())
+            # Check that the embedding layer and decoding layer are the same in size and in value
+            # self.assertTrue(check_same_values(embeddings, decoding))
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/llava_next/test_modeling_llava_next.py:376]: Opened https://github.com/huggingface/transformers/pull/29737 to fix this

File: src/transformers/utils/dummy_pt_objects.py [hunk 24]
```diff
@@ -4796,7 +4796,17 @@ def __init__(self, *args, **kwargs):
         requires_backends(self, ["torch"])
 
 
-class LlavaProcessor(metaclass=DummyObject):
```
Comment (ArthurZucker) [inline src/transformers/utils/dummy_pt_objects.py:4799]: I don't thin we should fully remove the LlavaProcessor no?

File: tests/models/musicgen_melody/test_modeling_musicgen_melody.py
Comment (ArthurZucker) [inline tests/models/musicgen_melody/test_modeling_musicgen_melody.py:1]: this is unrelated should be reverted

**Review by ArthurZucker — COMMENTED** [review]
LGTM let's just fix the last 2 nits!

File: src/transformers/utils/dummy_pt_objects.py (hunk 24 above)
Comment (NielsRogge) [reply src/transformers/utils/dummy_pt_objects.py:4799]: I think this is because LlavaProcessor was at the wrong place in the main init. Multimodel processors don't require a backend like torch.

File: src/transformers/models/llava/configuration_llava.py [hunk 25]
```diff
@@ -88,25 +88,36 @@ def __init__(
...
         self.image_token_index = image_token_index
         self.projector_hidden_act = projector_hidden_act
+
+        if "vocab_size" in kwargs and not isinstance(text_config, dict):
+            warnings.warn(
+                "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`.",
+                FutureWarning,
+            )
+            # set the vocab_size
+            text_config["vocab_size"] = kwargs["vocab_size"]
```
Comment (amyeroberts) [inline src/transformers/models/llava/configuration_llava.py:103]: No - this is still wrong and will break backwards compatibility

Comment (amyeroberts) [reply src/transformers/models/llava/configuration_llava.py:103]: vocab size from the kwargs was only ever set when text_config was `None`

File: tests/models/llava/test_modeling_llava.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/llava/test_modeling_llava.py:187]: That's because you removed the line that properly updates the config value in `resize_token_embeddings`

File: src/transformers/models/llava/configuration_llava.py (hunk 25 above)
Comment (NielsRogge) [reply src/transformers/models/llava/configuration_llava.py]: Just pushed an update which is resolves this. If not, let me know how we can make it backwards compatible.

File: tests/models/llava_next/test_modeling_llava_next.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/llava_next/test_modeling_llava_next.py:376]: rebase and make fix-copies should fix this now :)

File: tests/models/llava/test_modeling_llava.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/llava/test_modeling_llava.py:187]: Yes, I agree we don't want to just add stuff to make the tests pass. I mistakenly though `vocab_size` was still being set in the config for BC, in which case, we would have still needed to update the parameter. 

If we leave as a property to deprecate (cf notes in vip llava config file) then we can leave copied from headers here

File: src/transformers/models/vipllava/configuration_vipllava.py [hunk 26]
```diff
@@ -88,15 +87,21 @@ def __init__(
...
         self.vision_feature_layers = vision_feature_layers
-        self.vocab_size = vocab_size
+
+        if "vocab_size" in kwargs and not isinstance(text_config, dict):
+            warnings.warn(
+                "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`.",
+                FutureWarning,
+            )
+            # set the vocab_size
+            text_config["vocab_size"] = kwargs["vocab_size"]
```
Comment (amyeroberts) [inline src/transformers/models/vipllava/configuration_vipllava.py:104]: Accidentally left? This has the same BC issues as before and conflicts with the code below

```suggestion
```

File: src/transformers/models/llava/configuration_llava.py [hunk 27]
```diff
@@ -116,15 +119,22 @@ def __init__(
...
         elif text_config is None:
-            self.text_config = CONFIG_MAPPING["llama"]()
+            text_config = CONFIG_MAPPING["llama"]()
+            if "vocab_size" in kwargs:
+                warnings.warn(
+                    "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`.",
+                    FutureWarning,
+                )
+                # set the vocab_size
+                text_config.vocab_size = kwargs["vocab_size"]
```
Comment (amyeroberts) [inline src/transformers/models/llava/configuration_llava.py:136]: Same comments as for vip llava:
* Just have a warning at the start of the init
* Add some deprecation for `vocab_size`

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued iterations! 

Just a few final notes on BC for the config files

**Review by amyeroberts — APPROVED** [review]
Pushed the changes requested so all good to merge once tests pass!

**Comment by ArthurZucker** [comment]
https://github.com/huggingface/transformers/commit/a2a95164b82960ee1b36b01735b81a8a48ee1958 there were some unprotected imports 😈

File: src/transformers/models/llava/configuration_llava.py [hunk 28]
```diff
@@ -88,25 +88,34 @@ def __init__(
...
+
+        if "vocab_size" in kwargs:
+            warnings.warn(
+                "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`. Passing this argument has no effect",
+                FutureWarning,
+            )
+
         self.vision_feature_select_strategy = vision_feature_select_strategy
         self.vision_feature_layer = vision_feature_layer
-        self.vocab_size = vocab_size
```
Comment (fxmarty) [inline src/transformers/models/llava/configuration_llava.py:99]:
```
self = <tests.models.llava.test_modeling_llava.LlavaForConditionalGenerationIntegrationTest testMethod=test_small_model_integration_test_llama>

    @slow
    @require_bitsandbytes
    def test_small_model_integration_test_llama(self):
        # Let' s make sure we test the preprocessing to replace what is used
        model_id = "llava-hf/llava-1.5-7b-hf"
    
>       model = LlavaForConditionalGeneration.from_pretrained("llava-hf/llava-1.5-7b-hf", load_in_4bit=True)

tests/models/llava/test_modeling_llava.py:390: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/modeling_utils.py:3006: in from_pretrained
    config, model_kwargs = cls.config_class.from_pretrained(
src/transformers/configuration_utils.py:609: in from_pretrained
    return cls.from_dict(config_dict, **kwargs)
src/transformers/configuration_utils.py:761: in from_dict
    config = cls(**config_dict)
src/transformers/models/llava/configuration_llava.py:140: in __init__
    super().__init__(**kwargs)
src/transformers/configuration_utils.py:375: in __init__
    raise err
src/transformers/configuration_utils.py:372: in __init__
    setattr(self, key, value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = LlavaConfig {
  "architectures": [
    "LlavaForConditionalGeneration"
  ],
  "ignore_index": -100,
  "image_token_ind..._dim": 768,
    "vocab_size": 32000
  },
  "vision_feature_layer": -2,
  "vision_feature_select_strategy": "default"
}

key = 'vocab_size', value = 32064

    def __setattr__(self, key, value):
        if key in super().__getattribute__("attribute_map"):
            key = super().__getattribute__("attribute_map")[key]
>       super().__setattr__(key, value)
E       AttributeError: can't set attribute
```

Comment (fxmarty) [reply src/transformers/models/llava/configuration_llava.py:99]: will fix in https://github.com/huggingface/transformers/pull/29389

Comment (amyeroberts) [reply src/transformers/models/llava/configuration_llava.py:99]: sorry about this - thanks for addressing and fixing

**Comment by aliencaocao** [comment]
This model when loading gives `The model weights are not tied. Please use the `tie_weights` method before using the `infer_auto_device` function.`. I am using device_map='auto'. Anyway to fix this?

**Comment by amyeroberts** [comment]
Hi @aliencaocao, could you open an issue detailing this error and a minimal code snippet to reproduce? This will help us track what has and hasn't been addressed and help people find it if they're experiencing something similar.

**Comment by aliencaocao** [comment]
https://github.com/huggingface/transformers/issues/30001
